react/ComponentStyle.md

```diff
@@ -0,0 +1,304 @@
+    * UI do not use any utility function that has variables from outside.
+    * UI do not have children that are not PureComponent.
+
+* code snippet
+
+```
+// PureMadadata.jsx
+import React, { PureComponent, PropTypes } from 'react';
+
+class PureMadadata extends PureComponent {
+
+  constructor(props, context) {
+    super(props, context);
+    this.state = { value: 'madadata' };
+    this.handleSomethingClick = this.handleSomethingClick.bind(this);
+  }
+
+  componentWillMount() {
+    // * call this.setState here to update state, the next render will pick up state changes.
+    // * will only be invoked once.
+  }
+
+  componentDidMount() {
+    // * do setTimeout, setInterval, addEventListeners
+    // * do ajax calls (dispatch actions that will be handled by redux-saga)
+    // * access any refs to your children
+  }
+
+  componentWillReceiveProps(nextProps) {
+    // * componentWillReceiveProps enables you to decide what to do with next props before the component is rendered.
+    // * this.props -> old props
+    // * nextProps -> new props
+    // * will not be called for initial render
+    // * call this.setState here to update state, the next render will pick up state changes.
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    // * it will always return true by default. This method enables you to check to see whether you want react to re-render this component with new props and state. It's a costly operation and also a performance optimization point.
+    // * PureComponent will help you **shallow** compare the this.props and nextProps, this.state and nextState. If you need to do compare deeply, explicitly do this in the shouldComponentUpdate method of a Component instead of PureComponent.
+  }
+
+  componentWillUpdate(nextProps, nextState) {
+    // * componentWillUpdate enables you to perform preparation before and update occurs.
+    // * invoked immediately before rendering.
+    // * you might do:
+    //   1. set a variable for particular use
+    //   2. animate your DOM elements
+    //   3. dispatch action to do something to your store
+    //   ...
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    // * componentDidUpdate gives you a chance to operate on the DOM.
+    // * it will not be called in the initial render.
+  }
+
+  componentWillUnmount() {
+    // * perform necessary cleanup, clearTimer, clearEventListener ....
+  }
+
+  handleSomethingClick() {
+
+  }
+
+  render() {
+
+  }
+
+}
+
+PureMadadata.propTypes = {};
+PureMadadata.contextTypes = {};
+
+export default PureMadadata;
+
+```
+
+* Grammers
+  * if you don't do anything other than invoke the constructor of super class
+
+    ```
+    constructor() {
+      super();
+    }
+    ```
+    you can should omit the constructor.
+  * if you want to use props or both props and context in your constructor, you need to pass them to the constructor of its super class.
+
+    ```
+    constructor(props) {
+      super(props);
+      // this.props....
```

Comment:
I don't know. I guess there are identical.
But only after you super(props),
can you get props from **this** in constructor